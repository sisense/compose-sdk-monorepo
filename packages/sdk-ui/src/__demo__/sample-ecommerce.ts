import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
