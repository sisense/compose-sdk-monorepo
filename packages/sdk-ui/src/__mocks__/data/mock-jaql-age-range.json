{
  "headers": ["Age Range"],
  "metadata": [
    {
      "jaql": {
        "datatype": "text",
        "column": "Age Range",
        "merged": true,
        "dim": "[Commerce.Age Range]",
        "title": "Age Range",
        "table": "Commerce"
      },
      "instanceid": "859B4-BBF5-B8",
      "panel": "rows"
    }
  ],
  "datasource": {
    "revisionId": "TBD",
    "fullname": "LocalHost/Sample ECommerce"
  },
  "processingInfo": {
    "cacheType": "none",
    "cacheTime": "2023-12-07T13:54:29.289Z"
  },
  "translationInfo": {
    "translationServiceProvider": "NewTranslationService",
    "isTranslationFallback": false,
    "isQueryFallback": false,
    "translationDuration": 0.02,
    "sqlFromCache": false
  },
  "values": [
    [
      {
        "data": "0-18",
        "text": "0-18"
      }
    ],
    [
      {
        "data": "19-24",
        "text": "19-24"
      }
    ],
    [
      {
        "data": "25-34",
        "text": "25-34"
      }
    ],
    [
      {
        "data": "35-44",
        "text": "35-44"
      }
    ],
    [
      {
        "data": "45-54",
        "text": "45-54"
      }
    ],
    [
      {
        "data": "55-64",
        "text": "55-64"
      }
    ],
    [
      {
        "data": "65+",
        "text": "65+"
      }
    ]
  ]
}
