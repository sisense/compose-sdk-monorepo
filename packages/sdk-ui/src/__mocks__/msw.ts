import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import * as mockGlobals from '@/__mocks__/data/mock-globals.json';
import * as mockPalettes from '@/__mocks__/data/mock-palettes.json';

export const mockUrl = 'http://fake-url/';
export const mockToken = 'fake-token';
export const mockDashboardId = 'fake-dashboard-id';
export const mockWidgetId = 'fake-widget-id';

const handlers = [
  // Mock common calls made by SisenseContextProvider
  http.get('*/api/globals', () => HttpResponse.json(mockGlobals)),
  http.get('*/api/palettes/*', () => HttpResponse.json(mockPalettes)),

  // Mock tracking calls
  http.post('*/api/activities', () => new HttpResponse()),
];

export const server = setupServer(...handlers);
