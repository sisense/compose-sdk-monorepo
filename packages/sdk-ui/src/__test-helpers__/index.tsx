/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
