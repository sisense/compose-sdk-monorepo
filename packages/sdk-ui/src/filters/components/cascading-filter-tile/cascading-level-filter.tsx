import {
  DataSource,
  Filter,
  LevelAttribute,
  isMembersFilter,
  isDateRangeFilter,
  isRelativeDateFilter,
  isMeasureFilter,
  isTextFilter,
  isRankingFilter,
  isCustomFilter,
} from '@sisense/sdk-data';
import { MemberFilterTile } from '../member-filter-tile/index.js';
import { CriteriaFilterTile } from '../criteria-filter-tile/index.js';
import { DateRangeFilterTile, RelativeDateFilterTile } from '../date-filter/index.js';
import { CompleteFilterTileDesignOptions, FilterTile } from '../filter-tile.js';
import { useTranslation } from 'react-i18next';
import { useThemeContext } from '@/theme-provider';
import { TRIANGLE_COLOR_ADJUSTMENT, getSlightlyDifferentColor } from '@/utils/color';
import { CustomFilterTile } from '@/filters';

/**
 * Props of the {@link CascadingLevelFilterTile} component
 *
 * @internal
 */
export type CascadingLevelFilterTileProps = {
  /** Filter to display */
  filter: Filter;
  /**
   * Data source the query is run against - e.g. `Sample ECommerce`
   *
   * If not specified, the query will use the `defaultDataSource` specified in the parent Sisense Context.
   */
  dataSource?: DataSource;
  parentFilters: Filter[];
  /** Callback to handle filter change */
  onChange: (filter: Filter | null) => void;

  /** Whether the filter is the last in the list */
  isLast?: boolean;
};

const cascadingLevelTileDesign: CompleteFilterTileDesignOptions = {
  header: {
    shouldBeShown: true,
    isCollapsible: true,
    hasBorder: false,
    hasBackgroundFilterIcon: false,
  },
  border: {
    shouldBeShown: false,
  },
  footer: {
    shouldBeShown: false,
  },
} as const;

/**
 * Filter tile component that renders a level filter based on its type
 *
 * @internal
 */
export const CascadingLevelFilterTile = ({
  filter,
  dataSource,
  parentFilters,
  onChange,
  isLast,
}: CascadingLevelFilterTileProps) => {
  const { t } = useTranslation();
  const { themeSettings } = useThemeContext();
  const { backgroundColor: bgColor } = themeSettings.general;
  const triangleColor = getSlightlyDifferentColor(bgColor, TRIANGLE_COLOR_ADJUSTMENT);

  const attribute = filter.attribute;
  const title = attribute.name;
  const props = {
    attribute,
    dataSource,
    title,
    onChange,
    onUpdate: onChange,
    tileDesignOptions: cascadingLevelTileDesign,
  };

  const filterTile = isMembersFilter(filter) ? (
    <MemberFilterTile {...props} filter={filter} parentFilters={parentFilters} />
  ) : isDateRangeFilter(filter) ? (
    <DateRangeFilterTile {...props} filter={filter} attribute={attribute as LevelAttribute} tiled />
  ) : isRelativeDateFilter(filter) ? (
    <RelativeDateFilterTile {...props} filter={filter} arrangement="vertical" />
  ) : isMeasureFilter(filter) ||
    isMembersFilter(filter) ||
    isTextFilter(filter) ||
    isRankingFilter(filter) ? (
    <CriteriaFilterTile {...props} filter={filter} />
  ) : isCustomFilter(filter) ? (
    <CustomFilterTile {...props} filter={filter} />
  ) : (
    <FilterTile
      title={t('unsupportedFilter.title')}
      renderContent={() => (
        <p className="csdk-text-center csdk-text-[13px]">{t('unsupportedFilter.message')}</p>
      )}
      design={cascadingLevelTileDesign}
      locked={filter.locked}
    />
  );

  return (
    <div
      style={{ borderBottomColor: !isLast ? triangleColor : undefined }}
      className={!isLast ? `csdk-border-0 csdk-border-b csdk-border-solid` : ''}
    >
      {filterTile}
    </div>
  );
};
