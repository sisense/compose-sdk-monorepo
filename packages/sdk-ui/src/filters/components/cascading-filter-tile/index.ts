export { CascadingFilterTile, type CascadingFilterTileProps } from './cascading-filter-tile';
