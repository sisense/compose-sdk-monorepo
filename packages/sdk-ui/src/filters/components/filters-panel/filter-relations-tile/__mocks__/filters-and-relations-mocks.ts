import { FilterRelationsRules } from '@/utils/filter-relations';
import { Filter, filterFactory } from '@ethings-os/sdk-data';
import * as DM from '@/__test-helpers__/sample-ecommerce';

const countryFilter = filterFactory.members(DM.Country.Country, []);
const yearsFilter = filterFactory.members(DM.Commerce.Date.Years, []);
const revenueFilter = filterFactory.greaterThan(DM.Commerce.Revenue, 0);
const conditionFilter = filterFactory.members(DM.Commerce.Condition, ['Unspecified']);
const genderFilter = filterFactory.members(DM.Commerce.Gender, []);

export const filtersMock: Filter[] = [
  countryFilter,
  yearsFilter,
  revenueFilter,
  conditionFilter,
  genderFilter,
];

export const relationsMock: FilterRelationsRules = {
  left: {
    left: {
      instanceid: countryFilter.config.guid,
    },
    right: {
      instanceid: yearsFilter.config.guid,
    },
    operator: 'OR',
  },
  right: {
    left: {
      instanceid: revenueFilter.config.guid,
    },
    right: {
      left: {
        instanceid: conditionFilter.config.guid,
      },
      right: {
        instanceid: genderFilter.config.guid,
      },
      operator: 'AND',
    },
    operator: 'OR',
  },
  operator: 'AND',
};
