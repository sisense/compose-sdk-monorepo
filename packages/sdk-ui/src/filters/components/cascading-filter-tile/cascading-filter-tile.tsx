import { CascadingFilter, DataSource, Filter } from '@sisense/sdk-data';

import { useSynchronizedFilter } from '@/filters/hooks/use-synchronized-filter.js';
import { clearMembersFilter, cloneFilterAndToggleDisabled } from '@/utils/filters.js';

import { asSisenseComponent } from '../../../decorators/component-decorators/as-sisense-component.js';
import { FilterVariant } from '../common/filter-utils.js';
import { FilterTileContainer } from '../filter-tile-container.js';
import { CascadingLevelFilterTile } from './cascading-level-filter.js';

/**
 * Props for {@link CascadingFilterTile}
 *
 * @internal
 */
export interface CascadingFilterTileProps {
  /** Cascading filter object to initialize filter type and default values */
  filter: Filter;
  /**
   * Data source the query is run against - e.g. `Sample ECommerce`
   *
   * If not specified, the query will use the `defaultDataSource` specified in the parent Sisense Context.
   */
  dataSource?: DataSource;
  /** Arrangement of the filter inputs. Use vertical for standard filter tiles and horizontal for toolbars */
  arrangement?: FilterVariant;
  /** Callback returning filter object, or null for failure */
  onChange: (filter: Filter) => void;
  /** Filter delete callback */
  onDelete?: () => void;
  /** Filter edit callback */
  onEdit?: (levelIndex: number) => void;
}

/**
 * UI component representing group of cascading (dependent) filters.
 *
 * @param props - Cascading filter tile props
 * @returns Cascading filter tile component
 * @group Filter Tiles
 * @internal
 */
export const CascadingFilterTile = asSisenseComponent({ componentName: 'CascadingFilterTile' })(
  (props: CascadingFilterTileProps) => {
    const {
      filter: filterFromProps,
      arrangement,
      onChange: updateFilterFromProps,
      onDelete,
      dataSource,
      onEdit,
    } = props;

    const { filter, updateFilter } = useSynchronizedFilter<CascadingFilter>(
      filterFromProps as CascadingFilter,
      updateFilterFromProps,
    );

    const levelFilters = filter.filters;

    const handleLevelFilterChange = (
      changedLevelFilter: Filter,
      changedLevelFilterIndex: number,
    ) => {
      const newLevelFilters = filter.filters.map((levelFilter, index) => {
        if (index === changedLevelFilterIndex) {
          return changedLevelFilter;
        }
        if (index > changedLevelFilterIndex) {
          return clearMembersFilter(levelFilter);
        }
        return levelFilter;
      });

      const { guid, disabled } = filter.config;

      const newCascadingFilter = new CascadingFilter(newLevelFilters, { guid, disabled });

      updateFilter(newCascadingFilter);
    };

    const handleToggleDisabled = () => {
      const newCascadingFilter = cloneFilterAndToggleDisabled(filter);
      updateFilter(newCascadingFilter);
    };

    return (
      <FilterTileContainer
        renderContent={() => {
          const parentFilters: Filter[] = [];
          return levelFilters.map((levelFilter, index) => {
            const currentParentFilters = [...parentFilters];
            parentFilters.push(levelFilter);
            return (
              <CascadingLevelFilterTile
                key={index}
                filter={levelFilter}
                dataSource={dataSource}
                parentFilters={currentParentFilters}
                onChange={(newFilter) => {
                  return newFilter ? handleLevelFilterChange(newFilter, index) : null;
                }}
                isLast={index === levelFilters.length - 1}
                onEdit={onEdit ? () => onEdit?.(index) : undefined}
              />
            );
          });
        }}
        arrangement={arrangement}
        disabled={filter.config.disabled}
        onToggleDisabled={handleToggleDisabled}
        design={{
          header: {
            shouldBeShown: false,
            disableGroupHover: true,
          },
        }}
        locked={filter.config.locked}
        onDelete={onDelete}
      />
    );
  },
);
