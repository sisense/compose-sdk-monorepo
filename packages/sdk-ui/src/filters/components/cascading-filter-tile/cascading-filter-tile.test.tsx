import { render, screen } from '@testing-library/react';
import { http, HttpResponse } from 'msw';
import { CascadingFilter, Filter, filterFactory, isCascadingFilter } from '@sisense/sdk-data';
import { SisenseContextProviderProps } from '@/index';
import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import * as jaqlAgeRange from '@/__mocks__/data/mock-jaql-age-range.json';
import * as DM from '@/__test-helpers__/sample-ecommerce';
import { mockToken, mockUrl, server } from '@/__mocks__/msw';
import { CascadingFilterTile } from './cascading-filter-tile';
import { setTimeout } from 'timers/promises';

const contextProviderProps: SisenseContextProviderProps = {
  url: mockUrl,
  token: mockToken,
  defaultDataSource: 'Sample ECommerce',
  appConfig: {
    queryCacheConfig: { enabled: false },
    trackingConfig: { enabled: false },
  },
};

describe('CascadingFilterTile', () => {
  let cascadingFilter: Filter;
  beforeEach(() => {
    server.resetHandlers();
    server.use(
      http.post('*/api/datasources/:dataSource/jaql', () => HttpResponse.json(jaqlAgeRange)),
    );
    const topFilter = filterFactory.members(DM.Commerce.Gender, ['Unspecified']);
    const bottomFilter = filterFactory.members(DM.Commerce.AgeRange, ['0-18']);

    cascadingFilter = filterFactory.cascading([topFilter, bottomFilter]);
  });
  it('should render a CascadingFilterTile component', async () => {
    server.use(
      http.post('*/api/datasources/:dataSource/jaql', () => HttpResponse.json(jaqlAgeRange)),
    );

    if (!isCascadingFilter(cascadingFilter)) {
      throw new Error('Expected cascading filter');
    }

    render(
      <SisenseContextProvider {...contextProviderProps}>
        <CascadingFilterTile filter={cascadingFilter} onChange={() => {}} />
      </SisenseContextProvider>,
    );

    await setTimeout(200);

    await Promise.all(
      cascadingFilter.filters.map(async (filter) =>
        expect(await screen.findByText(filter.attribute.name)).toBeInTheDocument(),
      ),
    );
  });

  it('should render a locked CascadingFilterTile component', async () => {
    server.use(
      http.post('*/api/datasources/:dataSource/jaql', () => HttpResponse.json(jaqlAgeRange)),
    );

    const config = { locked: true };
    const topFilter = filterFactory.members(DM.Commerce.Gender, ['Unspecified'], {
      ...config,
      guid: 'member1',
    });
    const bottomFilter = filterFactory.members(DM.Commerce.AgeRange, ['0-18'], {
      ...config,
      guid: 'member2',
    });
    const lockedCascadingFilter = filterFactory.cascading([topFilter, bottomFilter], {
      ...config,
      guid: 'cascading1',
    });

    if (!isCascadingFilter(cascadingFilter)) {
      throw new Error('Expected cascading filter');
    }

    const { container } = render(
      <SisenseContextProvider {...contextProviderProps}>
        <CascadingFilterTile filter={lockedCascadingFilter} onChange={() => {}} />
      </SisenseContextProvider>,
    );

    await setTimeout(200);

    await Promise.all(
      cascadingFilter.filters.map(async (filter) =>
        expect(await screen.findByText(filter.attribute.name)).toBeInTheDocument(),
      ),
    );
    expect(container).toMatchSnapshot();
  });

  it('should be able to disable whole cascading filter', async () => {
    let updatedFilter: CascadingFilter | undefined;
    render(
      <SisenseContextProvider {...contextProviderProps}>
        <CascadingFilterTile
          filter={cascadingFilter}
          onChange={(newCascadingFilter) => {
            updatedFilter = newCascadingFilter as CascadingFilter;
          }}
        />
      </SisenseContextProvider>,
    );

    // disable filter with toggle
    const toggle = await screen.findByRole('switch');
    toggle.click();
    // check if cascading filter and all internal filters are disabled
    expect(updatedFilter && updatedFilter.config.disabled).toBe(true);
    updatedFilter!.filters.forEach((filter) => {
      expect(filter.config.disabled).toBe(true);
    });
  });
});
