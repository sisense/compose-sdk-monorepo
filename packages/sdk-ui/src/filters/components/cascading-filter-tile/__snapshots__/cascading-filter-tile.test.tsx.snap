// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CascadingFilterTile > should render a locked CascadingFilterTile component 1`] = `
<div>
  <div
    class="css-xhbwdz"
  >
    <div
      class="css-l4aa93"
      style="min-width: 216px; background-color: rgb(255, 255, 255);"
    >
      <main
        style="color: rgb(91, 99, 114); position: relative;"
      >
        <div
          class="csdk-border-0 csdk-border-b csdk-border-solid"
          style="border-bottom-color: rgba(232, 232, 232, 1);"
        >
          <div
            class="css-xhbwdz"
          >
            <div
              class="css-n759pn"
              style="min-width: 216px; background-color: rgb(255, 255, 255);"
            >
              <header
                class="css-1yqya5g"
                style="color: rgb(91, 99, 114);"
              >
                <svg
                  fill="#b4b4b4"
                  height="22"
                  viewBox="0 0 22 22"
                  width="22"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    transform="translate(5.000000,5.000000) scale(.70, 0.6)"
                  >
                    <path
                      d="m3,9v11h14V9M4,9V6c0-3.3 2.7-6 6-6c3.3,0 6,2.7 6,6v3H14V6c0-2.2-1.8-4-4-4-2.2,0-4,1.8-4,4v3"
                    />
                    <rect
                      fill="#dcdcdc"
                      height="9"
                      width="12"
                      x="4"
                      y="10"
                    />
                  </g>
                </svg>
                <span
                  class="csdk-text-[13px] csdk-mt-[6px] csdk-mb-[4px] csdk-ml-[7px] csdk-leading-[16px]"
                  style="color: rgb(91, 99, 114);"
                >
                  Gender
                </span>
              </header>
              <main
                style="color: rgb(91, 99, 114); position: relative;"
              >
                <div
                  class="csdk-flex csdk-flex-wrap csdk-p-3 csdk-gap-[5px]"
                >
                  <button
                    class="css-sc98s6"
                  >
                    Include all
                  </button>
                </div>
                <div
                  style="position: absolute; top: 4px; left: 4px; right: 4px; bottom: 4px; background-color: white; opacity: 0.5;"
                />
              </main>
            </div>
          </div>
        </div>
        <div
          class=""
        >
          <div
            class="css-xhbwdz"
          >
            <div
              class="css-n759pn"
              style="min-width: 214px; background-color: rgb(255, 255, 255);"
            >
              <div
                class="csdk-flex csdk-justify-center"
                data-testid="triangle-indicator"
              >
                <div
                  class="csdk-border-l-[70px] csdk-border-l-[transparent] csdk-border-r-[70px] csdk-border-r-[transparent] csdk-border-t-[13px]"
                  style="border-top-color: rgba(232, 232, 232, 1);"
                />
              </div>
              <header
                class="css-1yqya5g"
                style="color: rgb(91, 99, 114);"
              >
                <span
                  class="csdk-text-[13px] csdk-mt-[6px] csdk-mb-[4px] csdk-ml-[7px] csdk-leading-[16px]"
                  style="color: rgb(91, 99, 114);"
                >
                  AgeRange
                </span>
              </header>
              <main
                style="color: rgb(91, 99, 114); position: relative;"
              >
                <div
                  class="csdk-flex csdk-flex-wrap csdk-p-3 csdk-gap-[5px]"
                >
                  <button
                    class="css-sc98s6"
                  >
                    0-18
                  </button>
                </div>
                <div
                  style="position: absolute; top: 4px; left: 4px; right: 4px; bottom: 4px; background-color: white; opacity: 0.5;"
                />
              </main>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="css-1wpi5bx"
      />
    </div>
  </div>
</div>
`;
